### PR TITLE
bootstrap: use git merge-base for LLVM CI download logic

### LINKED ISSUE #101907: Merge commits break LLVM CI download
While PRs are not allowed to have merge commits, they can still be useful locally (or for not rewriting history for longer-running branches). 

However, they break the commit detection logic for downloading CI artifacts.

### Steps to Reproduce

1. `git checkout <old-commit>`
2. `git merge --no-ff master`
3. `./x.py build library/std`

### Expected Result

Builds fine.

### Actual Result

~~~
...
downloading https://ci-artifacts.rust-lang.org/rustc-builds/<old-commit>/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404      #                                                                               

error: failed to download llvm from ci

help: old builds get deleted after a certain time
help: if trying to compile an old commit of rustc, disable `download-ci-llvm` in config.toml:

[llvm]
download-ci-llvm = false

Build completed unsuccessfully in 0:01:43
~~~

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/101907
I tested this with a local branch that has extra merge commits due to Miri, and it worked fine there. But I am sure there are tons of other situations I did not think of...

r? @jyn514

### CODE REVIEW

**Comment by RalfJung** [comment]
Hm, this doesn't seem to work on CI already...
```
fatal: Not a valid object name origin/master
```
Does that checkout have the right remote under some other name? Or should we fall back to just `master` if we cannot find a suitable remote?

**Comment by jyn514** [comment]
don't have time for reviews

r? bootstrap

**Comment by RalfJung** [comment]
Looks like CI doesn't have any `master` branch of any sort, so the logic is bound to fail.

But I am confused that the logic even gets called there, doesn't CI need to always build LLVM itself anyway? The old logic would have returned the just-created fresh bors merge commit, and then it would have failed to find an LLVM build for that commit -- right?

**Comment by jyn514** [comment]
> But I am confused that the logic even gets called there, doesn't CI need to always build LLVM itself anyway? The old logic would have returned the just-created fresh bors merge commit, and then it would have failed to find an LLVM build for that commit -- right?

see the logic in `is_ci_llvm_modified`, which handles that case. maybe there's some way to unify it with the logic you're modifying, not sure.

cc https://github.com/rust-lang/rust/issues/113250

**Comment by albertlarsan68** [comment]
@rustbot author
Flip back to me once CI fixed :)

**Comment by albertlarsan68** [comment]
CI only runs when there are no merge conflicts, because it runs on the merge result.

**Comment by RalfJung** [comment]
Interesting, the rev-list path filtering seems to be doing absolutely nothing on CI. I guess this is because of the shallow clone.

I see two options for what we could do here:
- either we run something like `git fetch origin master --depth 1` on CI so that we know the current master branch, and can use it to determine the merge-base
- or we keep (some variant of) the old LLVM detection logic around as a fallback when `origin/master` cannot be found.

@albertlarsan68  any preferences?

**Comment by albertlarsan68** [comment]
I fyou are able to make option 1 work, the go ahead. Otherwise, I think the option 2 is a good compromise.

**Comment by RalfJung** [comment]
I couldn't quite figure out the best way to do this, so I went with a different approach: we just use `git merge-base` to get the starting point for the doing the "search last bors commit that changed LLVM" as before. That should also keep intact the assumptions made by `is_ci_llvm_modified`.

File: src/tools/build_helper/src/git.rs [hunk 1]
```diff
@@ -94,20 +103,13 @@ pub fn get_git_modified_files(
...
-    };
+    let merge_base = get_git_merge_base(git_dir)?;
 
-    let merge_base = output_result(git().arg("merge-base").arg(&updated_master).arg("HEAD"))?;
-    let files = output_result(git().arg("diff-index").arg("--name-only").arg(merge_base.trim()))?
+    let mut git = Command::new("git");
+    if let Some(git_dir) = git_dir {
+        git.current_dir(git_dir);
+    }
+    let files = output_result(git.arg("diff-index").arg("--name-only").arg(merge_base.trim()))?
```
Comment (albertlarsan68) [inline src/tools/build_helper/src/git.rs:112]:
```suggestion
    let files = output_result(git.args(["diff-index", "--name-only", merge_base.trim()]))?
```
Just personal preference, not blocking in any way

**Review by albertlarsan68 — APPROVED** [review]
LGTM, r=me with or without the change

File: src/tools/build_helper/src/git.rs (hunk 1 above)
Comment (RalfJung) [reply src/tools/build_helper/src/git.rs:112]: :shrug: this is existing code, I don't really care. However I see a lot of `.arg().arg()` chains throughout bootstrap so it doesn't seem like there is a consistent style either way.

**Pull request merged**